maitake/src/wait/dataqueue.rs

```diff
@@ -0,0 +1,984 @@
+    /// be linked into the list. If a node is known to not be linked, it is safe
+    /// to modify that node (such as by waking the stored [`Waker`]) without
+    /// holding the lock; otherwise, it may be modified through the list, so the
+    /// lock must be held when modifying the
+    /// node.
+    ///
+    /// A spinlock (from `mycelium_util`) is used here, in order to support
+    /// `no_std` platforms; when running `loom` tests, a `loom` mutex is used
+    /// instead to simulate the spinlock, because loom doesn't play nice with
+    /// real spinlocks.
+    queue: Mutex<List<Waiter<K, V>>>,
+
+    _key: PhantomData<K>,
+    _val: PhantomData<V>,
+}
+
+impl<K: PartialEq, V> Debug for DataWaitQueue<K, V> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("DataWaitQueue")
+            .field("state", &self.state)
+            .field("queue", &self.queue)
+            .field("key", &fmt::display(core::any::type_name::<K>()))
+            .field("val", &fmt::display(core::any::type_name::<V>()))
+            .finish()
+    }
+}
+
+/// Future returned from [`DataWaitQueue::wait()`].
+///
+/// This future is fused, so once it has completed, any future calls to poll
+/// will immediately return [`Poll::Ready`].
+#[derive(Debug)]
+#[pin_project(PinnedDrop)]
+#[must_use = "futures do nothing unless `.await`ed or `poll`ed"]
+pub struct Wait<'a, K: PartialEq, V> {
+    /// The [`DataWaitQueue`] being waited on from.
+    queue: &'a DataWaitQueue<K, V>,
+
+    /// Entry in the wait queue linked list.
+    #[pin]
+    waiter: Waiter<K, V>,
+}
+
+/// A waiter node which may be linked into a wait queue.
+#[repr(C)]
+#[pin_project]
+struct Waiter<K: PartialEq, V> {
+    /// The intrusive linked list node.
+    ///
+    /// This *must* be the first field in the struct in order for the `Linked`
+    /// implementation to be sound.
+    #[pin]
+    node: UnsafeCell<Node<K, V>>,
+
+    /// The future's state.
+    state: WaitStateBits,
+
+    key: K,
+    val: UnsafeCell<Option<V>>,
+}
+
+impl<K: PartialEq, V> Debug for Waiter<K, V> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("Waiter")
+            .field("node", &self.node)
+            .field("state", &self.state)
+            .field("key", &fmt::display(core::any::type_name::<K>()))
+            .field("val", &fmt::display(core::any::type_name::<V>()))
+            .finish()
+    }
+}
+
+#[repr(C)]
+struct Node<K: PartialEq, V> {
+    /// Intrusive linked list pointers.
+    ///
+    /// # Safety
+    ///
+    /// This *must* be the first field in the struct in order for the `Linked`
+    /// impl to be sound.
+    links: list::Links<Waiter<K, V>>,
+
+    /// The node's waker
+    waker: Wakeup,
+
+    // This type is !Unpin due to the heuristic from:
+    // <https://github.com/rust-lang/rust/pull/82834>
+    _pin: PhantomPinned,
+}
```

Comment:
this is identical to the `Node` in `WaitQueue`...it seems like we could just make this generic over the `Waiter` type, and reuse this code (so the waitmap would have a `Node<Waiter<K, V>>` while `WaitQueue` would have a `Node<Waiter>`).

Reply:
Seems reasonable!